File: problems/Course Schedule/canFinish.cs
public class Solution {
    public bool CanFinish(int numCourses, int[][] prerequisites) {
        var grid = new bool[numCourses, numCourses];
        var colours = new int[numCourses];

        foreach (var item in prerequisites) {
            grid[item[1], item[0]] = true;
        }

        for (var i = 0; numCourses > i; ++i) {
            for (var j = 0; numCourses > j; ++j) {
                if (grid[i, j] && 0 == colours[j] && !dfs(grid, colours, j)) {
                    return false;
                }
            }
        }

        return true;
    }

    private bool dfs(bool[,] grid, int[] colours, int idx = 0) {
        colours[idx] = 1;

        for (var i = 0; colours.Length > i; ++i) {
            if (grid[idx, i]) {
                if (1 == colours[i] || (0 == colours[i] && !dfs(grid, colours, i))) {
                    return false;
                }
            }
        }

        colours[idx] = 2;

        return true;
    }
}

File: problems/Course Schedule/canFinish.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<Integer>[] grid = new ArrayList[numCourses];
        int[] colours = new int[numCourses];

        for (int[] item : prerequisites) {
            if (null == grid[item[1]]) {
                grid[item[1]] = new ArrayList<>();
            }

            grid[item[1]].add(item[0]);
        }

        for (int idx = 0; numCourses > idx; ++idx) {
            if (!dfs(grid, colours, idx)) {
                return false;
            }
        }

        return true;
    }

    boolean dfs(List<Integer>[] grid, int[] colours, int idx) {
        colours[idx] = 1;

        if (null != grid[idx]) {
            for (int item : grid[idx]) {
                if (1 == colours[item] || (0 == colours[item] && !dfs(grid, colours, item))) {
                    return false;
                }
            }
        }

        colours[idx] = 2;

        return true;
    }
}
